pkg/servicemesh/controller/controller.go

```diff
@@ -121,19 +107,38 @@ func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(l
 		name:              name,
 	}
 
-	if smmrc.cacheWarmed {
-		smmrLog.Debugf("Listener for %q created. Ready to send an initial update", name)
+	// WARNING: This is a bit of a hack.  It is mostly intended as a temporary
+	// workaround to enable 2.1 testing to begin.
+	//
+	// Previously we sent an immediate initial update to all listeners when they
+	// were registered that included only the Istio system namespace.  That
+	// caused problems with some controllers, e.g. IOR, because they expect the
+	// callback to be called with the full authoritative set of namespaces and
+	// may remove resources for namespaces not in the list.
+	//
+	// This instead waits for the informer's cache to sync, then sends an
+	// initial update only if the expected SMMR is not found in the cache.
+	// Unfortunately, there is a race here between sending this initial update
+	// and a user creating an SMMR resource.
+	go func() {
+		// Ideally this would use the same stopCh as the controller, but
+		// ultimately it doesn't matter much.  If the channel is closed before
+		// the informer has synced this is all a no-op anyway.
+		stopCh := make(chan struct{})
+
+		if cache.WaitForNamedCacheSync("SMMR", stopCh, smmrc.informer.HasSynced) {
+			smmrLog.Infof("Cache synced for listener %q.", name)
+
+			_, exists, _ := smmrc.informer.GetStore().GetByKey(smmrc.namespace + "/" + smmrc.memberRollName)
+			if exists {
+				// No need to send initial update.  The informer will do it.
+				return
+			}
 
-		var members []string
-		for _, item := range smmrc.informer.GetIndexer().List() {
-			smmr := item.(*v1.ServiceMeshMemberRoll)
-			members = smmr.Status.ConfiguredMembers
+			smmrLog.Infof("Seeding listener %q with system namespace.", name)
```

Comment:
nit: the `updateNamespaces` func is already logging (at `debug` level), do we need this log line?

Reply:
Don't have a strong opinion, but this is only logged in the case where the SMMR doesn't exist. Seemed kind of nice to know without having to up the log level. Happy to make it debug or remove it though.